packages/flutter/lib/src/cupertino/colors.dart

```diff
@@ -80,3 +93,1101 @@ class CupertinoColors {
 }
+
+/// A [Color] subclass that represents a family of colors, and the currect effective
+/// color in the color family.
+///
+/// When used as a regular color, `CupertinoDynamicColor` is equivalent to the
+/// effective color (i.e. [CupertinoDynamicColor.value] will come from the effective
+/// color), which is determined by the [BuildContext] it is last resolved against. If it has never
+/// been resolved, typically the light, normal contrast, base elevation variant
+/// [CupertinoDynamicColor.color] will be the effective color.
+///
+// TODO(LongCatIsLooong): publicize once all Cupertino components have adopted this.
+// {@tool sample}
+//
+// The following snippet will create a [CupertinoButton] whose background color
+// is _lightModeColor_ in light mode but _darkModeColor_ in dark mode.
+//
+//
+// ```dart
+// CupertinoButton(
+//   child: child,
+//   color: CupertinoDynamicColor.withVibrancy(
+//     color: lightModeColor,
+//     darkColor: darkModeColor,
+//   ),
+//   onTap: () => null,
+// )
+// ```
+// {@end-tool}
+//
+// When a Cupertino component is provided with a `CupertinoDynamicColor`, either
+// directly in its constructor, or from an [InheritedWidget] it depends on (for example,
+// [DefaultTextStyle]), the component will automatically resolve the color by calling
+// [CupertinoDynamicColor.resolve], using their own [BuildContext].
+//
+// When used outside of a Cupertino component, such color resolution will not happen
+// automatically. It's essential to call [CupertinoDynamicColor.resolve] with the
+// correct [BuildContext] before using the color to paint, in order to get the
+// desired effect.
+///
+/// See also:
+///
+/// * [CupertinoUserInterfaceLevel], an [InheritedWidget] that may affect color
+/// resolution of a `CupertinoDynamicColor`.
+/// * [CupertinoSystemColors], an [InheritedWidget] that exposes system colors
+/// of iOS 13+.
+/// * https://developer.apple.com/documentation/uikit/uicolor/3238042-resolvedcolor.
+@immutable
+class CupertinoDynamicColor extends Color {
+  /// Creates an adaptive [Color] that changes its effective color based on the
+  /// [BuildContext] given. The default effective color is [color].
+  ///
+  /// All the colors must not be null.
+  CupertinoDynamicColor({
+    @required Color color,
+    @required Color darkColor,
+    @required Color highContrastColor,
+    @required Color darkHighContrastColor,
+    @required Color elevatedColor,
+    @required Color darkElevatedColor,
+    @required Color highContrastElevatedColor,
+    @required Color darkHighContrastElevatedColor,
+  }) : this._(
+         color,
+         <List<List<Color>>>[
+           <List<Color>>[
+             <Color>[darkColor, darkElevatedColor],
+             <Color>[darkHighContrastColor, darkHighContrastElevatedColor],
+           ],
+         <List<Color>>[
+             <Color>[color, elevatedColor],
+             <Color>[highContrastColor, highContrastElevatedColor],
+           ],
+         ],
+       );
```

Comment:
Can you explain this choice of data structure for _colorMap?  Is it just that you need to be able to group them and iterate them?

Reply:
It does seem that this structure adds a complication without much benefit that I can see (but there is a good chance I am missing something here).

Having a 3-dimensional list doesn't document itself very well. For example is the first index based on brightness, elevation or contrast? Does a dark brightness map to a 0 or a 1 index?

Perhaps a simpler approach would be to just treat the `CupertinoDynamicColor` as a data structure class with fields for each of the colors. The resolve methods would just choose which of the colors are needed based on the context properties. It would lead to a large if-then-else statement to pick which of the 8 colors it resolves to, but it would be explicit. I think it would also mean you wouldn't need the `_isPlatformBrightnessDependent` type functions as well.

Reply:
Yes, also in case in the future more traits are needed to resolve a dynamic color, or some traits become non-binary, it's much easier to maintain the resolve method this way, than having a bunch of `if` statements, or with bit acrobatics.

Reply:
updated the implementation of the getters, it should be more readable for the getters now.

Reply:
I think using a 3d list to represent the traits kinda mathematically makes sense, the 3 traits can be thought of as 3 orthogonal bases to form the whole trait space.

Some of the values are enums so it's worse than a few `if else`'s, it's going to be nested `switch`es, because the style guide discourages using `==` for enum matching.  

There're also some negligible optimizations in it:
- the `CupertinoDynamicColor.resolve` method is almost 0 copy because it just passes the reference of the array around.
- `CupertinoDynamicColor.==`is also faster (just realized this and made the change).